File: training/objectivec/CellarsAndCentaurs/Domain/Model/Dice.m
#import "Dice.h"

NSInteger rollD20(void) {
    DiceRoll *d20Roll = [[[Dice alloc] initWithSides:20] roll];
    return d20Roll.result;
}

@implementation DiceRoll

- (instancetype)initWithSides:(NSInteger)sides result:(NSInteger)result {
    self = [super init];
    if (self) {
        _sides = sides;
        _result = result;
    }
    return self;
}

@end

@implementation Dice

- (instancetype)initWithSides:(NSInteger)sides {
    self = [super init];
    if (self) {
        _sides = sides;
    }
    return self;
}

- (DiceRoll *)roll {
    NSInteger result = (NSInteger)arc4random_uniform((uint32_t)self.sides) + 1;
    return [[DiceRoll alloc] initWithSides:self.sides result:result];
}

@end

File: sample-projects/rust/src/domain/model/dice.rs
#[derive(Debug, Clone, Copy)]
pub struct Dice {
    pub sides: u32,
}

#[derive(Debug, Clone, Copy)]
pub struct DiceRoll {
    pub dice: Dice,
    pub result: u32,
}

impl Dice {
    pub fn roll(&self) -> DiceRoll {
        let d20_roll = (self.sides * 7 + 3) % self.sides + 1;
        DiceRoll {
            dice: *self,
            result: d20_roll,
        }
    }
}

pub fn roll_d20() -> DiceRoll {
    Dice { sides: 20 }.roll()
}
